drivers/hmis/app/models/hmis/hoh_change_handler.rb

```diff
@@ -0,0 +1,113 @@
+  # It provides methods to perform the change, apply the changes to the database, and validate the changes.
+  #
+  # This logic is split out from the UpdateRelationshipToHoH mutation in case we need to use it in cleanup routines,
+  # for example bulk assigning HoH to the oldest adult in a household.
+  #
+  # Parameters:
+  # - new_hoh_enrollment: The Enrollment that is becoming the Head of Household
+  # - hud_user_id: The "UserID" of the HUD user making the change.
+  class HohChangeHandler
+    attr_accessor :new_hoh_enrollment, :household_enrollments, :old_hoh_enrollments, :new_hoh_move_in_date, :old_hoh_move_in_date, :validation_errors, :hud_user_id
+
+    def initialize(new_hoh_enrollment:, hud_user_id:)
+      self.new_hoh_enrollment = new_hoh_enrollment
+      self.hud_user_id = hud_user_id
+
+      # All Enrollments in the household (including new and old HoHs)
+      self.household_enrollments = new_hoh_enrollment.household_members.preload(:client)
+      # Previous HoHs (should only be one, but we handle multiple for data cleanup)
+      self.old_hoh_enrollments = household_enrollments.sort_by(&:DateCreated).select(&:head_of_household?)
+
+      self.old_hoh_move_in_date = old_hoh_enrollments.map(&:move_in_date).compact.first
+      self.new_hoh_move_in_date = [old_hoh_move_in_date, new_hoh_enrollment.entry_date].max if old_hoh_move_in_date
+      self.validation_errors = HmisErrors::Errors.new
+    end
+
+    def validate(include_warnings: true)
+      return validation_errors unless include_warnings # there are no hard-stop validations at this time, to allow for data correction in tricky situations
+
+      old_hoh = old_hoh_enrollments.first&.client
+      # Add generic message indicating that HoH will change from X to Y
+      add_warning(self.class.change_hoh_message(old_hoh, new_hoh_enrollment.client))
+
+      # Add message about Move-in Date, only if the household Move-in Date is changing (which would occur if the new HoH entered after move-in)
+      add_warning(self.class.move_in_date_change_msg(old_hoh_move_in_date, new_hoh_move_in_date)) if new_hoh_move_in_date && new_hoh_move_in_date != old_hoh_move_in_date
+
+      # HoH shouldn't be WIP, unless all members are WIP
+      add_warning(self.class.incomplete_hoh_message) if new_hoh_enrollment.in_progress? && household_enrollments.not_in_progress.exists?
+
+      # HoH shouldn't be Exited, unless all clients are Exited
+      add_warning(self.class.exited_hoh_message) if new_hoh_enrollment.exit.present? && household_enrollments.open_on_date(Date.tomorrow).exists?
+
+      # HoH shouldn't be a child, unless all members are children
+      new_hoh_age = new_hoh_enrollment.client.age
+      add_warning(self.class.child_hoh_message) if new_hoh_age.present? && new_hoh_age < 18 && household_enrollments.find(&:adult?)
+
+      validation_errors
+    end
+
+    def apply_changes!
+      Hmis::Hud::Enrollment.transaction do
```

Comment:
I pushed a change like this https://github.com/greenriver/hmis-warehouse/pull/4852/commits/caf357b7b70a76fab78441e025785110cc0e8e2c :
```ruby
        Hmis::Hud::Enrollment.transaction do
          # Lock to avoid duplicate request collisions
          enrollment.household_members.lock!

          hoh_changer = Hmis::HohChangeHandler.new(new_hoh_enrollment: enrollment, hud_user_id: hmis_user.user_id)
          validations = hoh_changer.validate(include_warnings: !confirmed)
          return { errors: validations } if validations.any?

          hoh_changer.apply_changes!
        end
```
Thoughts? We only really need to lock the household members, not the entire project, which seems potentially disruptive. Of course there is the chance that an additional member could be added to the household, which this would miss, but that seems worth it to avoid locking the whole project imo. What do you think?



For my own reference, how I tested the behavior manually:
```ruby
# Process 1
Hmis::Hud::Enrollment.transaction do
  enrollment = Hmis::Hud::Enrollment.find(60298)
  enrollment.household_members.lock!

  enrollment.household_members.each do |member|
    member.move_in_date = Date.current
    member.save!
  end
  sleep 20
end

# Process 2 (waits and then fails with StaleObjectError)
Hmis::Hud::Enrollment.transaction do
  enrollment = Hmis::Hud::Enrollment.find(60298)
  enrollment.update!(move_in_date: 1.year.ago)
end

# Process 3 (doesn't wait, not part of household)
Hmis::Hud::Enrollment.transaction do
  enrollment = Hmis::Hud::Enrollment.find(51128)
  enrollment.update!(move_in_date: 1.year.ago)
end
```

Reply:
Thanks Gig

I would caution against locking more than one record (enrollment.household_members). The reason being that it might pose a risk of creating db deadlocks which can lead to app reliability issues. If you only lock one record, you avoid this issue completely. Note, if household_members is ordered by id, this becomes much less prone to deadlocks (but still a possibility).

I guess I'm not sure that locking the project would be disruptive in this case. The default lock is 'select for update' meaning that, while attempts to lock or update the project will block, reads are non-blocking. It seems like this operation should be quick so I don't know that this be much of an issue in practice but I might be missing something. I guess if there was some pathological case of a household containing thousands of enrollments this would be an issue.

Generally my thought is it would be nice to have some way to synchronize operations on an entire household and the project seems most natural for this, assuming we can keep our operations performant.

Happy to huddle about this to discuss further

Reply:
@ttoomey thanks for the feedback and explanation. Makes sense that it could deadlock, and that project locking shouldn't be disruptive provided that this operation is relatively fast (unless household size is huge like you say).
I adjusted  it to use `with_lock` on the project.